enhancements/update/to-multi-arch.md

```diff
@@ -0,0 +1,121 @@
+Because `ClusterOperator.status.versions` contains the information only about the operator's version number at the moment and `--to-multi-arch` starts the upgrade targeting *the same version*, the upgrade of each operator finishes right away from the CVO's point of view. Without the ability to recognize the component's turning into multi-arch, CVO's evaluation of the upgrade progress is inaccurate, i.e., the operators' upgrade is still ongoing when CVO claims complete. This enhancement is to define a proposal to address the issue.
+
+
+### User Stories
+
+* As a cluster admin, I want CVO to claim that migrating a cluster to multi-architecture is complete correctly.
+
+
+### Goals
+
+This enhancement is to provide a proposal that addresses the issue about the upgrade progress during a cluster migration to multi-arch. The proposal has impact only after the migration to multi-arch starts.
+
+### Non-Goals
+
+Whether or not 'oc adm upgrade' is executed succesfully or the requested upgrade is accepted is not in the scope of this enhancement. So it is _not_ required to have test cases for this enhancement to cover (although they are valid in general):
+- upgrade from multi-arch to single-arch,
+- upgrade from multi-arch to multi-arch, or
+- upgrade a cluster in a cloud provider where multi-arch is not supported to multi-arch.
+
+The `oc adm upgrade status` might be useful to monitor the migration progress to multi-arch but the output of the command does not block this enhancement and its implementation. The `status` command is designed for cluster upgrades but migration is not an upgrade conceptually and is implemented through an upgrade.
+
+## Proposal
+
+The proposal includes the steps that an OpenShift cluster operator may take to provide information about the upgrade in additional to the existing names and version numbers for the components that the operator manages.
+
+* Add a few elements into the existing list `ClusterOperator.status.versions` in the `ClusterOperator` manifest.
+Those new `versions` contains the image pull specifications for the operator and all the managed operands. Although the version stays the same, the image pull specification are different if a cluster becomes multi-arch.
+
+* Populate the `ClusterOperator.status.versions` including these new elements above.
+
+The CVO has already implemented the way that it checks if each element in the `ClusterVersion.status.versions` from the payload's has the counterpart in the actual one on the cluster.
+
+There is no need to extend the APIs in `openshift/api`, or modify the CVO implementation.
+
+
+### API Extensions
+
+None.
+
+### Topology Considerations
+
+TODO (hongkliu): I have no idea about how CVO works differently from Standalone Clusters.
```

Comment:
There's no CVO in MicroShift, so we're not relevant there.

HyperShift runs a CVO that watches ClusterOperators, just like in standalone.  Some of the component operators writing those ClusterOperator might run in the hosted cluster, just like standalone?  But some of them (network operator, etc.) run on the management cluster.  Even those still report state through the CVO via hosted-cluster-side ClusterOperator today, so this can work the same way in general.  Although there is no MCO on HyperShift; probably worth looping in the HyperShift folks to explain how the components that get ready to host multi-arch NodePools would report from HostedCluster.  [This][1] also landed a few weeks ago via openshift/hypershift#4538, and may bump up against this.

[1]: https://github.com/openshift/hypershift/blame/806c0519221383125c229147d4bd65bb90bcac80/api/hypershift/v1beta1/hostedcluster_types.go#L2786-L2790

Reply:
Updated with what I learned in the past few days.

> Although there is no MCO on HyperShift

[The docs](https://docs.openshift.com/container-platform/4.17/hosted_control_planes/hcp-updating.html#hcp-update-ocp-hc_hcp-updating) says there is a CVO for a hosted cluster does exist (just running on the management cluster?)

`rosa create 4.16.9 3h` made a hosted cluster and I see no CVO pod. Maybe the kubeconfig given by the cluster-bot do not have the permission to see it? It could be the same for MCO: It exists but I cannot see it because of permission?

If MCO does not exist anywhere for a hosted cluster, CVO would have troubles of reconcilation on `co/machine-config` as its `status.version` is not populated.

I did not go further on this because I was told that all COs for a standalone cluster get populated on a hosted cluster too. That is all we need for CVO to work for this enhancement, right?